### PR TITLE
feature/inf terraform azure

### PR DESCRIPTION
**Pull request by tbugfinder**
differences between aws and azure

https://github.com/nichtraunzer/ods-quickstarters/compare/ad0b199..e91d08a

Closes #...
Fixes #...

Tasks: 
- [x] Updated documentation in `docs/modules/...` directory
- [x] Ran tests in `<quickstarter>/testdata` directory

### CODE REVIEW

File: inf-terraform-azure/files/Makefile [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+
+.PHONY: install-python-env
+# Install python virtual environment based on Pipfile
+install-python-env:
+	CI=true PIPENV_VENV_IN_PROJECT=true pipenv install
+
+.PHONY: init-terraform
+# Install Terraform workspace.
+init-terraform:
+	echo 1 | terraform init -backend=true -force-copy -input=false -backend-config="backends/$(TF_BACKEND).tfbackend"
```
Comment (tbugfinder) [inline inf-terraform-azure/files/Makefile]: @matthiasbremser 
Please move the *.tfbackend files into folder environments and name those
[env].tfbackend.config

File: docs/modules/quickstarters/pages/inf-terraform-azure.adoc [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+This is a showcase to elaborate the combination of both technologies.
+
+The quickstarter includes kitchen-terraform with InSpec / cinc-auditor for testing.
+
+== What files / architecture is generated?
+
+----
+├── Jenkinsfile - This file contains Jenkins stages.
+├── README.md
+├── environments
```
Comment (tbugfinder) [inline docs/modules/quickstarters/pages/inf-terraform-azure.adoc:18]: @matthiasbremser 
please add docu about tfbackend config files

File: inf-terraform-azure/Jenkinsfile.template [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    stageTest(context)
+    stagePlan(context)
+    stageDeploy(context)
+    stageSmoke(context)
+    stageReport(context)
+}
+
+def getEnvSettings(def tfBackend, def workspace ) {
+  def config = [
+    "TF_BACKEND=${tfBackend}"
```
Comment (tbugfinder) [inline inf-terraform-azure/Jenkinsfile.template:46]: @matthiasbremser 
actually tfBackend is the environment in this case.
this is confusing

File: inf-terraform-azure/files/test/integration/default/controls/default.rb [hunk 4]
```diff
@@ -0,0 +1,21 @@
+require_relative '../libraries/terraform_data.rb'
+require_relative '../libraries/fixture_data.rb'
+require_relative '../libraries/azure.rb'
```
Comment (tbugfinder) [inline inf-terraform-azure/files/test/integration/default/controls/default.rb]: @matthiasbremser 
let's drop this - no ruby libraries, correct?

File: inf-terraform-azure/files/test/integration/default/inspec.yml [hunk 5]
```diff
@@ -0,0 +1,10 @@
+# This file has been created automatically.
+
+---
+name: stack
+supports:
+  - platform: azure
+depends:
+  - name: inspec-azure
+    git: https://github.com/inspec/inspec-azure
+    tag: v1.58.0
```
Comment (tbugfinder) [inline inf-terraform-azure/files/test/integration/default/inspec.yml]: @matthiasbremser bump version

File: inf-terraform-azure/files/test/integration/default/libraries/azure.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
+require 'azure_mgmt_authorization'
```
Comment (tbugfinder) [inline inf-terraform-azure/files/test/integration/default/libraries/azure.rb]: remove this file

File: inf-terraform-azure/Jenkinsfile.template (hunk 3 above)
Comment (matthiasbremser) [reply inf-terraform-azure/Jenkinsfile.template:46]: I suppose there are multiple environments: runtime environment, workspace (environment), backend (of a particular environment). Does this make sense?

File: inf-terraform-azure/files/Makefile [hunk 7]
```diff
@@ -133,10 +133,10 @@ smoke-test: install-test-deps
 	$(call check_defined,AZURE_CLIENT_ID)
 	$(call check_defined,AZURE_CLIENT_SECRET)
 
-	sh ./lib/scripts/createstackoutputs2yml.sh
+	bash ./lib/scripts/createstackoutputs2yml.sh
 	inspec_profiles=$$(ls -1 ./test/integration); \
 	for profile in $$inspec_profiles; do \
-		JSON_VARS_FILE="$(PWD)/terraform.tfvars.json" bundle exec cinc-auditor exec ./test/integration/$$profile --reporter=cli junit2:build/test-results/test/$$profile.xml json:reports/install/data/inspec/post-install/$$profile.json --target azure://; \
+		JSON_VARS_FILE="$(PWD)/terraform.tfvars.json" bundle exec cinc-auditor exec test/integration/$$profile --no-distinct-exit --reporter=cli json:reports/install/data/inspec/post-install/$$profile.json --input-file test/integration/default/files/inputs-from-tfo-stack.yml --target azure://; \
```
Comment (tbugfinder) [inline inf-terraform-azure/files/Makefile]: junit2 should not get dropped

File: docs/modules/quickstarters/pages/inf-terraform-azure.adoc [hunk 8]
```diff
@@ -0,0 +1,54 @@
+= Infrastructure Terraform Azure Quickstarter (inf-terraform-azure)
+
+The project supports generation of Azure deployments utilizing terraform tooling.
+installation and integration of it with OpenShift CD pipelines.
```
Comment (tbugfinder) [inline docs/modules/quickstarters/pages/inf-terraform-azure.adoc:4]: should it be .... terraform tooling (comma), installation and.....

File: docs/modules/quickstarters/pages/inf-terraform-azure.adoc [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+== Purpose of this quickstarter
+
+The quickstarter comes with a sample terraform resource which deploys nothing.
```
Comment (tbugfinder) [inline docs/modules/quickstarters/pages/inf-terraform-azure.adoc]: deploys a resource group and .....

File: docs/modules/quickstarters/pages/inf-terraform-azure.adoc [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+This is a showcase to elaborate the combination of both technologies.
+
+The quickstarter includes kitchen-terraform with InSpec / cinc-auditor for testing.
```
Comment (tbugfinder) [inline docs/modules/quickstarters/pages/inf-terraform-azure.adoc]: ....with Chef InSpec....

File: docs/modules/quickstarters/pages/inf-terraform-azure.adoc [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+* https://cinc.sh/start/auditor/[cinc-auditor]
+
+== Usage - how do you start after you provisioned this quickstarter
+
+Inject preferred tags into common-tags.tf, add/remove resources needed to main.tf.
+
+
+== How this quickstarter is built through Jenkins
+
+The Jenkinsfile is provisioned with this quick starter to ease CI/CD process. In Jenkinsfile, there are various stages.
```
Comment (tbugfinder) [inline docs/modules/quickstarters/pages/inf-terraform-azure.adoc]: .....this quickstarter....

File: inf-terraform-azure/Jenkinsfile [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+) { context ->
+
+  odsQuickstarterStageCopyFiles(context)
+
+  odsQuickstarterStageRenderJenkinsfile(context)
+
+  odsQuickstarterStageRenderJenkinsfile(
+    context,
+    [source: 'dev.yml.template',
+     target: 'environments/dev.yml']
```
Comment (tbugfinder) [inline inf-terraform-azure/Jenkinsfile:28]: this is confusing.
The PR includes environment/*.json files and docs about .json files.
However here we have *.ym templates.
Should *.yml files be described also ?

File: inf-terraform-azure/Jenkinsfile.template (hunk 3 above)
Comment (tbugfinder) [reply inf-terraform-azure/Jenkinsfile.template:46]: hm, the value of tfBackend, so I'm concerned about the variable name itself.
On the other hand this Jenkinsfile.template will be replaced.

File: inf-terraform-azure/README.md [hunk 13]
```diff
@@ -0,0 +1,6 @@
+# Azure Quickstarter (inf-terraform-azure)
+
+Documentation is located in our [official documentation](https://www.opendevstack.org/ods-documentation/ods-quickstarters/latest/index.html)
+
+Please update documentation in the [antora page directory](https://github.com/opendevstack/ods-quickstarters/tree/master/docs/modules/ROOT/pages)
```
Comment (tbugfinder) [inline inf-terraform-azure/README.md]: We have to update this file: ods-quickstarters/master/docs/modules/quickstarters/pages/index.adoc
It lacks also a toc entry for AWS.

Comment (tbugfinder) [inline inf-terraform-azure/README.md]: this link is invalid

File: inf-terraform-azure/files/.gitignore [hunk 14]
```diff
@@ -0,0 +1,17 @@
+.bundle
```
Comment (tbugfinder) [inline inf-terraform-azure/files/.gitignore:1]: please sync with 
https://github.com/opendevstack/ods-quickstarters/blob/master/inf-terraform-aws/files/.gitignore

File: inf-terraform-azure/files/.pre-commit-config.yaml [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+-   repo: https://github.com/PyCQA/flake8.git
+    rev: 4.0.1
+    hooks:
+    -   id: flake8
+        files: (\.py)$
+-   repo: https://github.com/antonbabenko/pre-commit-terraform
+    rev: v1.62.3
+    hooks:
+    -   id: terraform_fmt
+-   repo: ssh://git@bitbucket.biscrum.com:7999/infiaas/ods-pre-commit-hooks
```
Comment (tbugfinder) [inline inf-terraform-azure/files/.pre-commit-config.yaml]: this is not the correct url

File: inf-terraform-azure/files/.pre-commit-config.yaml [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    -   id: terraformcreatei2o
+        files: (\.tf)$
+    -   id: terraformstackmoduleoutputs
+        files: (\.tf)$
+-   repo: https://github.com/antonbabenko/pre-commit-terraform
+    rev: v1.62.3
+    hooks:
+    -   id: terraform_fmt
+    -   id: terraform_docs
+-   repo: ssh://git@bitbucket.biscrum.com:7999/infiaas/pre-commit-hooks-terraform.git
```
Comment (tbugfinder) [inline inf-terraform-azure/files/.pre-commit-config.yaml]: this is not the correct url

File: inf-terraform-azure/files/.pre-commit-config.yaml [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    rev: 4.0.1
+    hooks:
+    -   id: flake8
+        files: (\.py)$
+-   repo: https://github.com/antonbabenko/pre-commit-terraform
+    rev: v1.62.3
+    hooks:
+    -   id: terraform_fmt
+-   repo: ssh://git@bitbucket.biscrum.com:7999/infiaas/ods-pre-commit-hooks
+    rev: v0.3.6
```
Comment (tbugfinder) [inline inf-terraform-azure/files/.pre-commit-config.yaml]: v0.3.7

File: inf-terraform-azure/files/.ruby-version [hunk 18]
```diff
@@ -0,0 +1 @@
+2.7.4
```
Comment (tbugfinder) [inline inf-terraform-azure/files/.ruby-version]: 2.7.5

File: inf-terraform-azure/files/.terraform-version [hunk 19]
```diff
@@ -0,0 +1 @@
+1.0.3
```
Comment (tbugfinder) [inline inf-terraform-azure/files/.terraform-version]: 1.0.11

File: inf-terraform-azure/files/README.md [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+|------|
+| [local_file](https://registry.terraform.io/providers/hashicorp/local/latest/docs/resources/file) |
+| [random_id](https://registry.terraform.io/providers/hashicorp/random/3.1.0/docs/resources/id) |
+| [time_static](https://registry.terraform.io/providers/hashicorp/time/latest/docs/resources/static) |
+
+## Inputs
+
+| Name | Description | Type | Default | Required |
+|------|-------------|------|---------|:--------:|
+| data\_bucket\_name | The name of the S3 data bucket. | `string` | `"quickstarter"` | no |
```
Comment (tbugfinder) [inline inf-terraform-azure/files/README.md]: pre-commit hooks have to be executed.

File: inf-terraform-azure/files/README.md [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+| name | The name of the stack. |
+<!-- END OF PRE-COMMIT-TERRAFORM DOCS HOOK -->
+
+## How to test this Stack?
+
+
+Testing the functionality of this stack requires the following dependencies: `make`, `tee`, `ruby`, [`bundler`](https://bundler.io/), and [`terraform`](https://www.terraform.io/). Once installed, run `make test` from the command line.
+
+
+Note that, when running tests, stacks will interact with some cloud provider, such as *AWS*, *Azure* or *VMware*. It is up to you to provide sufficient configuration to enable these interactions, which differs between vendors. Here is an example for *Azure* that uses environment variables (via [Configuring the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html)):
```
Comment (tbugfinder) [inline inf-terraform-azure/files/README.md]: AWS Link?

File: inf-terraform-azure/files/README.md [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+$ export AZURE_CLIENT_ID=...
+$ export AZURE_CLIENT_SECRET=...
+$ make test
+```
+
+## How to extend a Stack?
+
+Extending a stack basically involves adding more blueprints whichever fit, and garnish it with custom infrastructure code if necessary.
+
+Setting up stack development guardrails requires the following dependencies: `make`, `tee`, `ruby`, [`bundler`](https://bundler.io/), `python`, [`pre-commit`](https://pre-commit.com/) [`terraform`](https://www.terraform.io/), and [`terraform-docs`](https://github.com/segmentio/terraform-docs). Once installed, run `make install-dev-deps` to install a set of quality improving *pre-commit hooks* into your local Git repository. Upon a `git commit`, these hooks will make sure that your code is both syntactically and functionally correct and that your `README.md` contains up-to-date documentation of your stack's supported set of *inputs* and *outputs*.
```
Comment (tbugfinder) [inline inf-terraform-azure/files/README.md]: there's an updated home (link) for terraform-docs. https://terraform-docs.io

File: inf-terraform-azure/files/README.md [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+
+## Environments
+The stack supports multiple environments (Testing/DEV/QA/PROD) within OpenDevStack. The behaviour of the stack in the environments can be controlled within the **environments** directory.
```
Comment (tbugfinder) [inline inf-terraform-azure/files/README.md]: The configuration files use
(slightly testing)
dev
test
prod

File: inf-terraform-azure/files/README.md [hunk 24]
```diff
@@ -0,0 +1,87 @@
...
+The *.yml files define the Jenkins secrets to read and are used to deploy into the right environments.
+The *.json files can override variables from **variables.tf** in case different environments request different inputs (e.g. deploy a smaller version of the stack in DEV).
```
Comment (tbugfinder) [inline inf-terraform-azure/files/README.md]: smaller version -> do we talk about a different sizing of resources?

File: inf-terraform-azure/files/Makefile [hunk 25]
```diff
@@ -0,0 +1,242 @@
...
+install-test-deps: install-ruby-gems install-python-env
+	# output azure account and user id for testing
+	az login --service-principal -u $(AZURE_CLIENT_ID) -p $(AZURE_CLIENT_SECRET) --tenant $(AZURE_TENANT_ID)
+	az account show --output tsv | tee $(INSTALL_REPORT_HOME)/azure_deploy_account.log
+
+.PHONY: cinc-auditor-test
+# run cinc-auditor without use of kitchen-terraform and create yaml for mapping terraform outputs to inspec inputs.
+cinc-auditor-test:
+	sh ./lib/scripts/createstackfixtureoutputs2yml.sh
+	bundle exec cinc-auditor exec test/integration/default --no-create-lockfile --no-distinct-exit --input-file ./test/integration/$$profile/files/inputs-from-tfo-stack.yml --target azure://
```
Comment (tbugfinder) [inline inf-terraform-azure/files/Makefile]: $$profile -> please replace by default

File: inf-terraform-azure/files/Makefile [hunk 26]
```diff
@@ -0,0 +1,242 @@
...
+# Checks if a variable is defined and produces a fatal error if not.
+check_defined = \
+	$(strip $(foreach 1,$1, \
+		$(call __check_defined,$1,$(strip $(value 2)))))
+__check_defined = \
+	$(if $(value $1),, \
+		$(error Undefined $1$(if $2, ($2))$(if $(value @), \
+			required by target `$@')))
+
+# Execute tests in included blueprints.
```
Comment (tbugfinder) [inline inf-terraform-azure/files/Makefile]: anything following from here was removed

compare to https://github.com/opendevstack/ods-quickstarters/blob/master/inf-terraform-aws/files/Makefile#L156-L162

File: inf-terraform-azure/files/main.tf [hunk 27]
```diff
@@ -0,0 +1,23 @@
...
+    DeploymentDate        = formatdate("YYYYMMDD", timestamp())
+    InitialDeploymentDate = time_static.deployment.rfc3339
+  })
+}
+
+resource "time_static" "deployment" {}
+
+resource "azurerm_resource_group" "this" {
+  name     = local.unique_name
+  location = local.location
```
Comment (tbugfinder) [inline inf-terraform-azure/files/main.tf:15]: add tags?

File: inf-terraform-azure/files/main.tf [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+}
+
+resource "azurerm_resource_group_template_deployment" "this" {
+  name                = local.unique_name
+  resource_group_name = azurerm_resource_group.this.name
+  deployment_mode     = "Incremental"
+  template_content    = file("${path.module}/arm-templates/storage-account.json")
```
Comment (tbugfinder) [inline inf-terraform-azure/files/main.tf:24]: add tags

File: inf-terraform-azure/files/metadata.yml [hunk 29]
```diff
@@ -0,0 +1,7 @@
+---
+name: AZURE
+# yamllint disable-line rule:line-length
+description: "Azure offers reliable, scalable, and inexpensive cloud computing services. Free to join, pay only for what you use."
```
Comment (tbugfinder) [inline inf-terraform-azure/files/metadata.yml:4]: obviously - thanks for clarification.

File: inf-terraform-azure/files/test/fixtures/default/random.tf [hunk 30]
```diff
@@ -0,0 +1,19 @@
+provider "random" {}
+
+resource "random_id" "id" {
+  byte_length = 4
+}
+
+resource "random_string" "passworddb" {
```
Comment (tbugfinder) [inline inf-terraform-azure/files/test/fixtures/default/random.tf:7]: which db?

File: inf-terraform-azure/files/test/integration/default/controls/default.rb [hunk 31]
```diff
@@ -0,0 +1,20 @@
...
+f = SpecHelper::FixtureData.new.for_module(name)
+
+control 'stack' do
+  impact 1.0
+  title  "Test Suite: 'Stack'"
+  desc   "This test suite asserts the correct functionality of the stack under test."
+  tag    name
+
+  describe "Stack Testing" do
+    it { expect(true).to be_truthy }
```
Comment (tbugfinder) [inline inf-terraform-azure/files/test/integration/default/controls/default.rb]: we should include use of the generic resource test.

File: inf-terraform-azure/files/variables.tf [hunk 32]
```diff
@@ -0,0 +1,39 @@
...
+# -----------------------------------------------------------------------------
+
+variable "name" {
+  description = "The name of the stack."
+  type        = string
+  default     = "stack-azure-quickstarter-delete-me"
+}
+
+variable "is_test" {
+  description = "Whether is is part of a test execution or not. Defaults to false."
```
Comment (tbugfinder) [inline inf-terraform-azure/files/variables.tf]: whether it is ....

File: inf-terraform-azure/files/versions.tf [hunk 33]
```diff
@@ -0,0 +1,21 @@
+terraform {
+  required_version = ">= 0.13"
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = "2.95.0"
```
Comment (tbugfinder) [inline inf-terraform-azure/files/versions.tf]: I'm proposing 3.0.2

File: inf-terraform-azure/files/versions.tf [hunk 34]
```diff
@@ -0,0 +1,21 @@
+terraform {
+  required_version = ">= 0.13"
```
Comment (tbugfinder) [inline inf-terraform-azure/files/versions.tf]: >=1.0.0

File: inf-terraform-azure/Jenkinsfile (hunk 12 above)
Comment (matthiasbremser) [reply inf-terraform-azure/Jenkinsfile:28]: Added explanations in docs/modules/quickstarters/pages/inf-terraform-azure.adoc

**Comment by matthiasbremser** [comment]
### Default Test Suite

Profile: stack
Version: (not specified)
Target:  azure://d4d824c7-6c3b-4b08-a272-2e36112eb706

  ✔  stack: Test Suite: 'Stack'
     ✔  Azure Generic Resource - api_version: 2022-01-01 latest: /subscriptions/d4d824c7-6c3b-4b08-a272-2e36112eb706/resourceGroups/stack-azure-quickstarter-test-4f97c53f is expected to exist
     ✔  Azure Generic Resource - api_version: 2021-04-01 latest: /subscriptions/d4d824c7-6c3b-4b08-a272-2e36112eb706/resourceGroups/stack-azure-quickstarter-test-4f97c53f/providers/Microsoft.Resources/deployments/stack-azure-quickstarter-test-4f97c53f is expected to exist


Profile: Azure Resource Pack (inspec-azure)
Version: 1.86.0
Target:  azure://d4d824c7-6c3b-4b08-a272-2e36112eb706

     No tests executed.

Profile Summary: 1 successful control, 0 control failures, 0 controls skipped
Test Summary: 2 successful, 0 failures, 0 skipped

**Pull request closed without merging**